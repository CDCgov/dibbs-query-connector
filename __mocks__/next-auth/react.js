const useSession = jest
  .fn()
  .mockReturnValue({ data: undefined, status: "loading" });

const signOut = jest.fn();
/**
 *
 * @param root0 - root0
 * @param root0.children - root0.children
 * @returns - JSX.Element
 */
const SessionProvider = ({ children }) => <>{children}</>;

export { useSession, signOut, SessionProvider };
