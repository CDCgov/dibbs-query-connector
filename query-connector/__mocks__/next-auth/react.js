/* eslint-disable jsdoc/require-returns, jsdoc/require-param-description */
// __mocks__/next-auth/react.js
const useSession = jest.fn().mockReturnValue({ data: undefined });

/**
 *
 * @param root0
 * @param root0.children
 */
const SessionProvider = ({ children }) => <>{children}</>;

export { useSession, SessionProvider };
