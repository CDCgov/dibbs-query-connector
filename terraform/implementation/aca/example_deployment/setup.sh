#!/bin/bash

# If you're running this code from an M-series processor or other ARM variant, you may need to specify the container platform manually.
export DOCKER_DEFAULT_PLATFORM=linux/amd64

# This section provides an example of targeted deployments, if required for your installation.
terraform apply -target=module.foundations -target=module.networking
terraform apply -target=module.container_apps