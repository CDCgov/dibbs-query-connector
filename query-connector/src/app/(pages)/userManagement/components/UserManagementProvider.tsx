"use client";

import { createContext, useState } from "react";
import { showToastConfirmation } from "@/app/ui/designSystem/toast/Toast";
import { User } from "@/app/models/entities/users";
import { QueryTableResult } from "../../queryBuilding/utils";

export type SubjectType = "Members" | "Queries" | null;

interface UserManagementData {
  teamQueryEditSection: {
    isOpen: boolean;
    title: string;
    subtitle: string;
    placeholder: string;
    groupId: string;
    subjectType: SubjectType;
    subjectData: User[] | QueryTableResult[];
  };
}

interface UserManagementContext extends UserManagementData {
  openEditSection: (
    title: string,
    subtitle: string,
    subjectType: SubjectType,
    groupId: string,
    subjectData?: User[] | QueryTableResult[],
  ) => void;
  closeEditSection: () => void;
  handleSearch: (searchFilter: string) => void;
  handleQueryUpdate: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const initData: UserManagementData = {
  teamQueryEditSection: {
    isOpen: false,
    title: "",
    subtitle: "",
    placeholder: "Search",
    groupId: "",
    subjectType: null,
    subjectData: [],
  },
};

export const UserManagementContext = createContext<UserManagementContext>({
  ...initData,
  openEditSection: () => {},
  closeEditSection: () => {},
  handleSearch: (_searchFilter: string) => {},
  handleQueryUpdate: () => {},
});

/**
 *
 * @param root0 - DataProvider props (only receives children)
 * @param root0.children - React components that need access to the user management data
 * @returns DataProvider component which allows to access and update user management data
 */
const DataProvider: React.FC<React.PropsWithChildren> = ({ children }) => {
  const [innerState, setInnerState] = useState<UserManagementData>(initData);

  /**
   * Event handlers for edit section
   */
  function closeEditSection() {
    setInnerState(initData);
  }

  function openEditSection(
    title: string,
    subtitle: string,
    subjectType: SubjectType,
    id: string,
    subjectData?: User[] | QueryTableResult[],
  ) {
    let placeholder = "";

    if (subjectType == "Members") {
      placeholder = "Search members";
    } else {
      placeholder = "Search queries";
    }
    const newData = subjectData ?? [];
    const newState: UserManagementData = {
      ...innerState,
      teamQueryEditSection: {
        ...innerState.teamQueryEditSection,
        title,
        subtitle,
        placeholder,
        subjectType,
        groupId: id,
        subjectData: newData,
        isOpen: true,
      },
    };
    setInnerState(newState);
  }

  function handleSearch(filter: string) {
    // TODO data filtering
    console.log("filtering ...", filter);
  }

  async function handleQueryUpdate(e: React.ChangeEvent<HTMLInputElement>) {
    const queryName = e.currentTarget.name;
    const checked = e.target.checked;

    const alertText = checked
      ? `Assigned ${queryName} to ${innerState.teamQueryEditSection.title}`
      : `Removed ${queryName} from ${innerState.teamQueryEditSection.title}`;

    try {
      showToastConfirmation({
        body: alertText,
      });
    } catch (error) {
      showToastConfirmation({
        heading: "Something went wrong",
        body: alertText,
        variant: "error",
      });
      console.error("Error updating group queries list:", error);
    }
  }

  return (
    <UserManagementContext.Provider
      value={{
        ...innerState,
        openEditSection,
        closeEditSection,
        handleSearch,
        handleQueryUpdate,
      }}
    >
      {children}
    </UserManagementContext.Provider>
  );
};

export default DataProvider;
