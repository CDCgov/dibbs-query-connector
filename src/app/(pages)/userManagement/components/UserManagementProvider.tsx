"use client";

import { createContext, useState } from "react";
import { User } from "@/app/models/entities/users";
import { CustomUserQuery } from "@/app/models/entities/query";

export type SubjectType = "Members" | "Queries" | null;

interface UserManagementData {
  teamQueryEditSection: {
    isOpen: boolean;
    title: string;
    subtitle: string;
    placeholder: string;
    groupId: string;
    subjectType: SubjectType;
    subjectData: User[] | CustomUserQuery[];
  };
}

interface UserManagementContext extends UserManagementData {
  openEditSection: (
    title: string,
    subtitle: string,
    subjectType: SubjectType,
    groupId: string,
    subjectData: User[] | CustomUserQuery[],
  ) => void;
  closeEditSection: () => void;
}

const initData: UserManagementData = {
  teamQueryEditSection: {
    isOpen: false,
    title: "",
    subtitle: "",
    placeholder: "Search",
    groupId: "",
    subjectType: null,
    subjectData: [],
  },
};

export const UserManagementContext = createContext<UserManagementContext>({
  ...initData,
  openEditSection: () => {},
  closeEditSection: () => {},
});

/**
 *
 * @param root0 - DataProvider props (only receives children)
 * @param root0.children - React components that need access to the user management data
 * @returns DataProvider component which allows to access and update user management data
 */
const DataProvider: React.FC<React.PropsWithChildren> = ({ children }) => {
  const [innerState, setInnerState] = useState<UserManagementData>(initData);

  /**
   * Event handlers for edit section
   */
  function closeEditSection() {
    setInnerState(initData);
  }

  function openEditSection(
    title: string,
    subtitle: string,
    subjectType: SubjectType,
    id: string,
    subjectData: User[] | CustomUserQuery[],
  ) {
    let placeholder = "";

    if (subjectType == "Members") {
      placeholder = "Search members";
    } else {
      placeholder = "Search queries";
    }

    const newState: UserManagementData = {
      ...innerState,
      teamQueryEditSection: {
        ...innerState.teamQueryEditSection,
        title,
        subtitle,
        placeholder,
        subjectType,
        groupId: id,
        subjectData,
        isOpen: true,
      },
    };
    setInnerState(newState);
  }

  return (
    <UserManagementContext.Provider
      value={{
        ...innerState,
        openEditSection,
        closeEditSection,
      }}
    >
      {children}
    </UserManagementContext.Provider>
  );
};

export default DataProvider;
