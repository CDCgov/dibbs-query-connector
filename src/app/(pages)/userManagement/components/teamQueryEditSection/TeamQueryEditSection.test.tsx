import React, { ReactNode } from "react";
import { waitFor, screen, render } from "@testing-library/react";
import {
  allGroups,
  mockAdmin,
  mockGroupWithSingleQuery,
  mockQueryNoGroups,
  mockQueryWithGroups,
  mockStandard,
  mockSuperAdmin,
} from "../../test-utils";
import { RootProviderMock } from "@/app/tests/unit/setup";
import { SubjectType, UserManagementContext } from "../UserManagementProvider";
import UserManagementDrawer from "./TeamQueryEditSection";

jest.mock("next-auth/react");

jest.mock(
  "@/app/ui/components/withAuth/WithAuth",
  () =>
    ({ children }: React.PropsWithChildren) => <div>{children}</div>,
);

jest.mock("@/app/backend/dbServices/queryBuilding/service", () => ({
  getCustomQueries: jest.fn(),
}));

jest.mock("@/app/backend/user-management", () => ({
  getGroupMembers: jest.fn().mockResolvedValue({ items: [], totalItems: 0 }),
  getGroupQueries: jest.fn().mockResolvedValue({ items: [], totalItems: 0 }),
  getUsers: jest.fn().mockResolvedValue({ items: [], totalItems: 0 }),
  getUserGroups: jest.fn().mockResolvedValue({ items: [], totalItems: 0 }),
}));

describe("User Management drawer", () => {
  const allQueries = [mockQueryNoGroups, mockQueryWithGroups];
  const allUsers = [mockAdmin, mockStandard, mockSuperAdmin];

  const renderOpenDrawer = (subject: SubjectType, children: ReactNode) => {
    return (
      <UserManagementContext.Provider
        value={{
          teamQueryEditSection: {
            isOpen: true,
            title: mockGroupWithSingleQuery.name,
            subtitle: subject == "Queries" ? "Assigned queries" : "Members",
            placeholder: "Search",
            groupId: mockGroupWithSingleQuery.id,
            subjectType: subject,
            subjectData: subject == "Queries" ? allQueries : allUsers,
          },
          openEditSection: jest.fn(),
          closeEditSection: jest.fn(),
        }}
      >
        {children}
      </UserManagementContext.Provider>
    );
  };
  it("is closed by default", async () => {
    render(
      <RootProviderMock currentPage="/userManagement">
        <UserManagementDrawer
          userGroups={[]}
          setUserGroups={jest.fn()}
          users={[]}
          setUsers={jest.fn()}
          refreshView={jest.fn()}
          activeTabLabel={""}
          allQueries={[]}
          setAllQueries={jest.fn()}
        />
      </RootProviderMock>,
    );

    const drawer = screen.getByTestId("drawer-open-false");
    expect(drawer.classList).toContain("closed");
    expect(document.body).toMatchSnapshot();

    jest.restoreAllMocks();
  });
  it("renders an error message if there is no subject data to display", async () => {
    render(
      <RootProviderMock currentPage="/userManagement">
        {renderOpenDrawer(
          "Members",
          <UserManagementDrawer
            userGroups={[]}
            setUserGroups={jest.fn()}
            users={[]}
            setUsers={jest.fn()}
            refreshView={jest.fn()}
            activeTabLabel={""}
            allQueries={[]}
            setAllQueries={jest.fn()}
          />,
        )}
      </RootProviderMock>,
    );

    const drawer = screen.getByTestId("drawer-open-true");
    expect(drawer.classList).toContain("open");

    const body = document.getElementsByClassName("drawerBody")[0];
    expect(body.innerHTML).toContain("No users found");
    expect(document.body).toMatchSnapshot();

    jest.restoreAllMocks();
  });
  it("renders the list of all available queries", async () => {
    render(
      <RootProviderMock currentPage="/userManagement">
        {renderOpenDrawer(
          "Queries",
          <UserManagementDrawer
            userGroups={allGroups}
            setUserGroups={jest.fn()}
            users={allUsers}
            setUsers={jest.fn()}
            refreshView={jest.fn()}
            activeTabLabel={""}
            allQueries={allQueries}
            setAllQueries={jest.fn()}
          />,
        )}
      </RootProviderMock>,
    );

    await waitFor(() => {
      expect(screen.queryByText("No queries found")).not.toBeInTheDocument();
    });

    const drawer = screen.getByTestId("drawer-open-true");
    expect(drawer.classList).toContain("open");

    expect(screen.getAllByRole("listitem")).toHaveLength(allQueries.length);
    expect(document.body).toMatchSnapshot();

    jest.restoreAllMocks();
  });

  it("renders the list of all available users", async () => {
    render(
      <RootProviderMock currentPage="/userManagement">
        {renderOpenDrawer(
          "Members",
          <UserManagementDrawer
            userGroups={allGroups}
            setUserGroups={jest.fn()}
            users={allUsers}
            setUsers={jest.fn()}
            refreshView={jest.fn()}
            activeTabLabel={""}
            allQueries={allQueries}
            setAllQueries={jest.fn()}
          />,
        )}
      </RootProviderMock>,
    );

    await waitFor(() => {
      expect(screen.queryByText("No users found")).not.toBeInTheDocument();
    });

    const drawer = screen.getByTestId("drawer-open-true");
    expect(drawer.classList).toContain("open");

    expect(screen.getAllByRole("listitem")).toHaveLength(allUsers.length);
    expect(document.body).toMatchSnapshot();

    jest.restoreAllMocks();
  });
});
