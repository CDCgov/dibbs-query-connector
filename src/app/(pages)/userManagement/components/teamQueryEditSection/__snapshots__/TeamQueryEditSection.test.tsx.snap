// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`User Management drawer is closed by default 1`] = `
<body>
  <div>
    <div>
      <div
        aria-label="drawer-container"
        class="drawer closed width35"
        data-testid="drawer-open-false"
        id="drawer-container"
        role="dialog"
      >
        <div
          class="drawerContent display-none"
        >
          <div
            class="drawerHeader"
          >
            <button
              aria-label="Close drawer"
              class="closeButton"
              data-testid="close-drawer"
            >
              <svg
                aria-label="X icon indicating closure"
                class="usa-icon usa-icon--size-3"
                focusable="false"
                height="1em"
                role="img"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </button>
            <h2
              class="margin-0 padding-bottom-2"
              data-testid="drawer-title"
              id="drawer-title"
            />
            <div>
              <input
                class="usa-input searchField"
                data-testid="textInput"
                id="searchFieldTemplate"
                name="searchFieldTemplate"
                placeholder="Search"
                type="search"
                value=""
              />
            </div>
          </div>
          <div
            class="drawerBody"
          >
            <div
              class="margin-top-2"
            >
              No queries found.
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`User Management drawer renders an error message if there is no subject data to display 1`] = `
<body>
  <div>
    <div>
      <div
        aria-label="drawer-container"
        class="drawer open width35"
        data-testid="drawer-open-true"
        id="drawer-container"
        role="dialog"
      >
        <div
          class="drawerContent display-block"
        >
          <div
            class="drawerHeader"
          >
            <button
              aria-label="Close drawer"
              class="closeButton"
              data-testid="close-drawer"
            >
              <svg
                aria-label="X icon indicating closure"
                class="usa-icon usa-icon--size-3"
                focusable="false"
                height="1em"
                role="img"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </button>
            <h2
              class="margin-0 padding-bottom-0"
              data-testid="drawer-title"
              id="drawer-title"
            >
              S.P.E.W.
            </h2>
            <h3
              class="subtitle"
            >
              Members
            </h3>
            <div>
              <input
                class="usa-input searchField"
                data-testid="textInput"
                id="searchFieldTemplate"
                name="searchFieldTemplate"
                placeholder="Search"
                type="search"
                value=""
              />
            </div>
          </div>
          <div
            class="drawerBody"
          >
            <div
              class="margin-top-2"
            >
              No users found.
            </div>
          </div>
        </div>
      </div>
      <div
        class="overlay"
      />
    </div>
  </div>
</body>
`;

exports[`User Management drawer renders the list of all available queries 1`] = `
<body>
  <div>
    <div>
      <div
        aria-label="drawer-container"
        class="drawer open width35"
        data-testid="drawer-open-true"
        id="drawer-container"
        role="dialog"
      >
        <div
          class="drawerContent display-block"
        >
          <div
            class="drawerHeader"
          >
            <button
              aria-label="Close drawer"
              class="closeButton"
              data-testid="close-drawer"
            >
              <svg
                aria-label="X icon indicating closure"
                class="usa-icon usa-icon--size-3"
                focusable="false"
                height="1em"
                role="img"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </button>
            <h2
              class="margin-0 padding-bottom-0"
              data-testid="drawer-title"
              id="drawer-title"
            >
              S.P.E.W.
            </h2>
            <h3
              class="subtitle"
            >
              Assigned queries
            </h3>
            <div>
              <input
                class="usa-input searchField"
                data-testid="textInput"
                id="searchFieldTemplate"
                name="searchFieldTemplate"
                placeholder="Search"
                type="search"
                value=""
              />
            </div>
          </div>
          <div
            class="drawerBody"
          >
            <ul
              aria-description="queries for S.P.E.W."
              class="usa-list--unstyled margin-top-2"
            >
              <li>
                <div
                  class="usa-checkbox margin-bottom-3 checkbox"
                  data-testid="checkbox"
                >
                  <input
                    checked=""
                    class="usa-checkbox__input"
                    id="q1"
                    name="Test Query 1"
                    type="checkbox"
                  />
                  <label
                    class="usa-checkbox__label"
                    for="q1"
                  >
                    Test Query 1
                  </label>
                </div>
              </li>
              <li>
                <div
                  class="usa-checkbox margin-bottom-3 checkbox"
                  data-testid="checkbox"
                >
                  <input
                    checked=""
                    class="usa-checkbox__input"
                    id="q2"
                    name="Test Query 2"
                    type="checkbox"
                  />
                  <label
                    class="usa-checkbox__label"
                    for="q2"
                  >
                    Test Query 2
                  </label>
                </div>
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        class="overlay"
      />
    </div>
  </div>
</body>
`;

exports[`User Management drawer renders the list of all available users 1`] = `
<body>
  <div>
    <div>
      <div
        aria-label="drawer-container"
        class="drawer open width35"
        data-testid="drawer-open-true"
        id="drawer-container"
        role="dialog"
      >
        <div
          class="drawerContent display-block"
        >
          <div
            class="drawerHeader"
          >
            <button
              aria-label="Close drawer"
              class="closeButton"
              data-testid="close-drawer"
            >
              <svg
                aria-label="X icon indicating closure"
                class="usa-icon usa-icon--size-3"
                focusable="false"
                height="1em"
                role="img"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </button>
            <h2
              class="margin-0 padding-bottom-0"
              data-testid="drawer-title"
              id="drawer-title"
            >
              S.P.E.W.
            </h2>
            <h3
              class="subtitle"
            >
              Members
            </h3>
            <div>
              <input
                class="usa-input searchField"
                data-testid="textInput"
                id="searchFieldTemplate"
                name="searchFieldTemplate"
                placeholder="Search"
                type="search"
                value=""
              />
            </div>
          </div>
          <div
            class="drawerBody"
          >
            <ul
              aria-description="members of S.P.E.W."
              class="usa-list--unstyled margin-top-2"
            >
              <li>
                <div
                  class="usa-checkbox margin-bottom-3 checkbox"
                  data-testid="checkbox"
                >
                  <input
                    class="usa-checkbox__input"
                    id="456"
                    name="lilypotter"
                    type="checkbox"
                  />
                  <label
                    class="usa-checkbox__label"
                    for="456"
                  >
                    Lily Potter
                  </label>
                </div>
              </li>
              <li>
                <div
                  class="usa-checkbox margin-bottom-3 checkbox"
                  data-testid="checkbox"
                >
                  <input
                    class="usa-checkbox__input"
                    id="789"
                    name="hermionegranger"
                    type="checkbox"
                  />
                  <label
                    class="usa-checkbox__label"
                    for="789"
                  >
                    Hermione Granger
                  </label>
                </div>
              </li>
              <li>
                <div
                  class="usa-checkbox margin-bottom-3 checkbox"
                  data-testid="checkbox"
                >
                  <input
                    class="usa-checkbox__input"
                    id="123"
                    name="harrypotter"
                    type="checkbox"
                  />
                  <label
                    class="usa-checkbox__label"
                    for="123"
                  >
                    Harry Potter
                  </label>
                </div>
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        class="overlay"
      />
    </div>
  </div>
</body>
`;
